drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1901 @@
+
+	/* Advance read FIFO */
+	ep_cur->read_fifo_pos += *size_p;
+	ep_cur->read_fifo_used -= *size_p;
+	if (ep_cur->read_fifo_used == 0) {
+		ep_cur->read_fifo_pos = ep_cur->dmabuf_base;
+	}
+
+	return 0;
+}
+
+/* Copy data from user buffer to EP FIFO
+ *
+ * size_p holds size to copy/copied on input/output
+ */
+static int numaker_usbd_ep_fifo_copy_from_user(struct numaker_usbd_ep *ep_cur,
+					       const uint8_t *usrbuf, uint32_t *size_p)
+{
+	__ASSERT_NO_MSG(size_p);
+	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
+	__ASSERT_NO_MSG(ep_cur->ep_mps_valid);
+	__ASSERT_NO_MSG(ep_cur->ep_mps <= ep_cur->dmabuf_size);
+
+	*size_p = Z_MIN(*size_p,
+			numaker_usbd_ep_fifo_max(ep_cur) - numaker_usbd_ep_fifo_used(ep_cur));
+
+	uint32_t dmabuf_addr = USBD_BUF_BASE + USBD_GET_EP_BUF_ADDR(ep_cur->usbd_hw_ep_hndl);
+
+	USBD_MemCopy((uint8_t *)dmabuf_addr, (uint8_t *)usrbuf, *size_p);
+
+	/* Advance write FIFO */
+	ep_cur->write_fifo_pos += *size_p;
+	ep_cur->write_fifo_free -= *size_p;
+	if (ep_cur->write_fifo_free == 0) {
+		ep_cur->write_fifo_pos = ep_cur->dmabuf_base;
+	}
+
+	return 0;
+}
+
+/* Update EP read/write FIFO on DATA OUT/IN completed */
+static void numaker_usbd_ep_fifo_update(struct numaker_usbd_ep *ep_cur)
+{
+	__ASSERT_NO_MSG(ep_cur->ep_addr_valid);
+	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
+
+	if (USB_EP_DIR_IS_OUT(ep_cur->ep_addr)) {
+		/* Read FIFO */
+		/* NOTE: For one-shot implementation, FIFO gets updated from empty. */
+		ep_cur->read_fifo_pos = ep_cur->dmabuf_base;
+		/* NOTE: See comment on mxpld_ctrlout for why make one copy of CTRL OUT's MXPLD */
+		if (USB_EP_GET_IDX(ep_cur->ep_addr) == 0) {
+			ep_cur->read_fifo_used = ep_cur->mxpld_ctrlout;
+		} else {
+			ep_cur->read_fifo_used = USBD_GET_PAYLOAD_LEN(ep_cur->usbd_hw_ep_hndl);
+		}
+	} else {
+		/* Write FIFO */
+		/* NOTE: For one-shot implementation, FIFO gets to empty. */
+		ep_cur->write_fifo_pos = ep_cur->dmabuf_base;
+		ep_cur->write_fifo_free = numaker_usbd_ep_fifo_max(ep_cur);
+	}
+}
+
+/* Configure EP DMA buffer */
+static void numaker_usbd_ep_config_dmabuf(struct numaker_usbd_ep *ep_cur, uint32_t dmabuf_base,
+					  uint32_t dmabuf_size)
+{
+	USBD_SET_EP_BUF_ADDR(ep_cur->usbd_hw_ep_hndl, dmabuf_base);
+
+	ep_cur->dmabuf_valid = true;
+	ep_cur->dmabuf_base = dmabuf_base;
+	ep_cur->dmabuf_size = dmabuf_size;
+
+	/* FIFO is implemented based on DMA buffer, which needs reset accordingly. */
+	numaker_usbd_ep_fifo_reset(ep_cur);
+}
+
+/* Abort EP */
+static void numaker_usbd_ep_abort(struct numaker_usbd_ep *ep_cur)
+{
+	struct usb_dc_numaker_device *dev = ep_cur->dev;
+	const struct usb_dc_numaker_config *config = dev->config;
+
+	/* Abort EP on-going transaction */
+	USBD_STOP_TRANSACTION(ep_cur->usbd_hw_ep_hndl);
+
+#if defined(CONFIG_USB_DC_NUMAKER_WORKAROUND_DISALLOW_ISO_IN_OUT_SAME_NUM)
+	if (ep_cur->usbd_hw_ep_cfg & USBD_CFG_TYPE_ISO) {
+		/* Restore to not enabling the function */
+		*((volatile uint32_t *)(config->usbd_base->RESERVE0 + 0)) |= 0x2;
```

Comment:
maybe add a define for that `0x02`, also writing to a register called `RESERVE0` is a bit odd, should that be renamed as well?

Reply:
still relevant?

Reply:
this workaround relies on hidden h/w register, which is undocumented in soc spec, so the code is intended to be opaque

Reply:
this workaround relies on hidden h/w register, though, it is undocumented in soc spec, so the code here is intended to be opaque

Reply:
renamed this bit and register with macro